lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -36,6 +39,89 @@ K_SEM_DEFINE(trace_done_sem, 1, 1);
 static int trace_init(void);
 static int trace_deinit(void);
 
+#if CONFIG_NRF_MODEM_LIB_TRACE_BITRATE
+static double min_bps = DBL_MAX;
+static double max_bps;
+static double tot_bps;
+static int num_bps;
+static int64_t bitrate_start;
+static int64_t bitrate_end;
+
+#if CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG
+#define LOG_BITRATE_PERIOD K_MSEC(CONFIG_NRF_MODEM_LIB_TRACE_BITRATE_LOG_PERIOD_MS)
+
+static void log_bitrate_stats(struct k_work *item);
+
+K_WORK_DELAYABLE_DEFINE(log_bitrate_stats_work, log_bitrate_stats);
+
+static void log_bitrate_stats(struct k_work *item)
+{
+	LOG_INF("Trace bitrate (bps): %10.2lf", tot_bps / num_bps);
+
+	k_work_schedule(&log_bitrate_stats_work,
+			LOG_BITRATE_PERIOD);
+}
+#endif
+
+static void update_bps(double bps)
+{
+	if (bps < min_bps) {
+		min_bps = bps;
+	}
+
+	if (bps > max_bps) {
+		max_bps = bps;
+	}
+
+	tot_bps += bps;
+	num_bps++;
+}
+
+int nrf_modem_lib_trace_backend_bitrate_get(double *min_val, double *max_val, double *avg_val)
+{
+	if (min_val == NULL || max_val == NULL || avg_val == NULL) {
+		return -EINVAL;
+	}
+
+	*min_val = min_bps;
+	*max_val = max_bps;
+	*avg_val = tot_bps / num_bps;
+
+	return 0;
+}
+
+static void trace_backend_bitrate_perf_start(void)
+{
+	bitrate_start = k_uptime_ticks();
+}
+
+static void trace_backend_bitrate_perf_end(int size)
+{
+	double delta;
+	double bps;
+	double time;
+
+	bitrate_end = k_uptime_ticks();
+
+	if (size > 0) {
+		/* Casting `int64_t` to `double` */
+		delta = bitrate_end - bitrate_start;
+		time = delta / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+		bps = size * 8 / time;
```

Comment:
I think this would be simplified by using `k_uptime_get`/`k_uptime_get_32`, give it a look

Reply:
Not precise enough.